Tximport in Salmon Pipeline

dongbohu opened the pull request:
## Issue Number
#82 
#243
#273

## Purpose/Implementation Notes
This PR is much larger than I expected. It includes:
- tximport data analysis in `Salmon` pipeline
- a unit test of `tximport`
- updated docker images for salmon pipeline
- #273 is fixed as a byproduct. 

## Types of changes
- new features
- Bugfix (non-breaking change which fixes an issue)

## Functional tests
Tested by the unit test in this PR

## Checklist

_Put an `x` in the boxes that apply._

- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

dongbohu commented:
@jaclyn-taroni: As you can see, I modified your R script a little bit. The main changes are:
- Add logic to handle the two arguments
- Changed the way to find `quant.sf` files and the corresponding sample names (due to the structure of files generated by the download and `salmon quant` step.

jaclyn-taroni commented on workers/data_refinery_workers/processors/tximport.R:
```diff
@@ -0,0 +1,63 @@
+# Generate the gene to transcript mapping based on the input file
+# specified by "--gene2txmap" option:
+gene2tx.file <- opt$gene2txmap
+gene2tx.df <- readr::read_tsv(gene2tx.file, col_names = c("gene_id", "tx_name", "DROP"))
```
The way we're doing this now, there shouldn't be trailing tabs. So I think this should do it:
```R
gene2tx_df <- readr::read_tsv(gene2tx_file, col_names = c("gene_id", "tx_name))
```
And you won't need line 40.

dongbohu replied:
Sure.

jaclyn-taroni commented on workers/data_refinery_workers/processors/tximport.R:
```diff
@@ -0,0 +1,63 @@
+# (2) TPM file (saved as <experiment_dir>/gene_lengthScaledTPM.tsv)
+
+# Handle input arguments:
+library("optparse")
```
Where possible, I think we should go the [double colon route](https://stackoverflow.com/questions/23232791/is-it-a-good-practice-to-call-functions-in-a-package-via) (e.g., `optparse::`) rather than loading/attaching packages (to avoid namespace collisions). But it's only one library, so probably not a big deal. 

Below `make_option()` would become `optparse::make_option()`, `optparse::OptionParser`, etc.

jaclyn-taroni commented on workers/data_refinery_workers/processors/tximport.R:
```diff
@@ -0,0 +1,63 @@
+opt <- parse_args(opt_parser)
+
+# directory of the experiment data
+exp.directory <- opt$exp_dir
```
I think we should keep the style consistent and have this become `exp_directory`

jaclyn-taroni commented on workers/data_refinery_workers/processors/tximport.R:
```diff
@@ -0,0 +1,63 @@
+# Get a list of "quant.sf" files in experiment directory.
+# The path of each "quant.sf" are in this format:
+# <exp_dir>/<sample>/processed/quant.sf
+sf.files <- list.files(exp.directory, pattern="^quant.sf$", full.names = TRUE, recursive = TRUE)
```
Same thing here, if we want to make this more pythonic style wise `sf_files`

jaclyn-taroni commented on workers/data_refinery_workers/processors/tximport.R:
```diff
@@ -0,0 +1,63 @@
+# Run tximport and save RDS
+txi.out <- tximport::tximport(files = sf.files,
+                              type = "salmon",
+                              txOut = TRUE,
```
Based on the discussion over on #82, I think this should be `txOut  = FALSE`. (I would probably change the variable name to `txi` in that case.) Doing that & saving the RDS would mean we have:

* gene-level TPM
* gene-level counts
* gene-level offset [appropriate for use with differential expression analysis](http://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html#use-with-downstream-bioconductor-dge-packages)

The way we have it here gives us [transcript-level estimates as a list of matrices](http://bioconductor.org/packages/release/bioc/vignettes/tximport/inst/doc/tximport.html#import-transcript-level-estimates), which we could extract from the `quants.sf` file should folks want that.

dongbohu replied:
Sure.

jaclyn-taroni commented on workers/data_refinery_workers/processors/tximport.R:
```diff
@@ -0,0 +1,63 @@
+
+# Run tximport::summarizeToGene() and save length-scaled
+# TPM (transcripts per million) file.
+txi.sum <- tximport::summarizeToGene(txi = txi.out,
```
If we make the above change, I think it makes more sense to change this to
```R
txi_length_scaled <- tximport::tximport(files = sf_files,
                                        type = "salmon", 
                                        tx2gene = tx2gene,
                                        countsFromAbundance = "lengthScaledTPM")
```

dongbohu replied:
Sure.

jaclyn-taroni commented on workers/data_refinery_workers/processors/tximport.R:
```diff
@@ -0,0 +1,63 @@
+                                     countsFromAbundance = "lengthScaledTPM")
+# As data.frame, write to file
+lstpm.df <- as.data.frame(txi.sum$counts)
+lstpm.df <- cbind(rownames(lstpm.df), lstpm.df)
```
If there will be [tidyverse](https://www.tidyverse.org) packages (like `readr`, which is used above) on the Docker image, I believe we can convert lines 60-61 to

```R
lstpm_df <- tibble::rownames_to_column(df, var = "Gene")
```

dongbohu replied:
Sounds good.

dongbohu replied:
You mean: 
```lstpm_df <- tibble::rownames_to_column(lstpm_df, var = "Gene")```
right? (since the first argument `df` is an undefined variable)

dongbohu commented:
By the way, don't worry about the `All checks have failed` warning. I cancelled the CircleCI tests to save some resource for Kurt and Rich, because this PR has nothing to do with the other tests (yet).

jaclyn-taroni commented:
Changes LGTM 👍 , thanks @dongbohu !

dongbohu commented on workers/data_refinery_workers/processors/tximport.R:
```diff
@@ -0,0 +1,63 @@
+# This script is based on the one written by J. Taroni:
```
Jackie has reviewed this file. So please feel free to skip this one.

dongbohu commented on workers/dockerfiles/Dockerfile.salmon:
```diff
@@ -23,41 +30,29 @@ WORKDIR /home/user
-
 RUN tar -xzf Salmon-${SALMON_VERSION}_linux_x86_64.tar.gz
-
-# Salmon can extract to a different directory than the name of the tar file.
```
The main reason for this change: when the pre-compiled salmon library files were copied into /usr/local/lib/, they have higher priorities than the library files in `/usr/lib/`, and messed up both JRE and Rscript. Alternatively, we can compile the source code of `salmon`, but they requires boost library, which makes the docker at least 150MB image larger. So I tweaked the PATH (see below) instead.

kurtwheeler replied:
Wait I'm confused, I don't actually see you tweaking the PATH at all. It looks like you're making `/usr/local/bin` into a softlink which points to the `bin` that came with salmon? It seems like if you wanted to add salmon's bin to the PATH you could just append it to the `$PATH` environment variable right? It also seems like another potential solution could be change `$PATH` to prioritize `/usr/lib` over `/usr/local/lib`?

No matter how we end up doing this, I do think the explanation should be a comment in the code instead of on the PR so that we can see it if we ever go to modify this later.

dongbohu replied:
Originally I played with the $PATH, then I realized that no matter $PATH is defined in ~/.bashrc or ~/.profile, it will be effective only in BASH, so adding a soft link in `/usr/local/bin/` seems better. (Generally it is not a good idea to tweak the dynamic library path, because it may affect other programs.)

dongbohu replied:
I will add a comment line in this file.

dongbohu commented on workers/dockerfiles/Dockerfile.salmon:
```diff
@@ -23,41 +30,29 @@ WORKDIR /home/user
+RUN rm -f Salmon-${SALMON_VERSION}_linux_x86_64.tar.gz
 # End Salmon installation.
 
-# Package list based on:
```
I removed these packages because these packages were required to compile salmon, and we are using pre-compiled version. (As a matter of fact, most of these packages have already been installed implicitly because they were the dependencies of `r-base`.)

dongbohu commented on workers/dockerfiles/Dockerfile.salmon:
```diff
@@ -23,41 +30,29 @@ WORKDIR /home/user
 # We need JRE for QC tools
-RUN apt-get update -y
-RUN apt-get install -y unzip default-jre --fix-missing
+RUN apt-get update -y && apt-get install -y default-jre --fix-missing
```
I removed `unzip` here because it is already installed implicitly as a dependency of `r-base`.

dongbohu commented on workers/dockerfiles/Dockerfile.salmon:
```diff
@@ -69,7 +64,7 @@ COPY common/dist/data-refinery-common-* common/
 
 USER user
-
+RUN echo "export PATH=`pwd`/Salmon-latest_linux_x86_64/bin:$PATH" >> ~/.bashrc
```
I didn't put it in `~/.profile` or `~/.bash_profile` because we would like the new PATH to be effective in all bash sessions (not only in BASH login shell).

dongbohu commented:
I added some extra comments in the PR to (hopefully) make the review easier.

dongbohu commented:
@jaclyn-taroni @kurtwheeler and @Miserlou: This PR is ready for review now! Thanks.

kurtwheeler commented on workers/data_refinery_workers/processors/salmon.py:
```diff
@@ -373,6 +396,58 @@ def _run_fastqc(job_context: Dict) -> Dict:
+    # TODO: Because this job is done at experiment level, it doesn't make
+    # sense to update job_context (which is at sample level right now).
+    # The DB model may have to be modified to associate this ComputationalResult
+    # record with the experiment (instead of samples in the experiment).
```
I disagree that we should modify the DB model. There's a many to many relationship between Samples and ComputationalResults for exactly this use case. I think every Sample from this experiment should be connected to this result.

dongbohu replied:
One reason why I don't like this idea is because when a sample belongs to multiple experiments, then a sample will be associated with multiple tximport computational results. This seems confusing.

kurtwheeler replied:
Hmm, that's a good point. I'm not sure which I like less. Maybe @Miserlou has thoughts on this?

kurtwheeler replied:
I think we need to connect it to SOMETHING, as it is here these ComputationalResults can't even be queried for except by the `command_executed` field. I'm cool with picking whichever option is easier and going with that, opening an issue about it, and figuring out the best solution later, but I think we need to be able to find the `tximport` ComputationalResult for a given sample/experiment for these to be useful at all.

dongbohu replied:
Unless we change the DB schema, the only table that  it can associate with right now is `SampleResultAssociate`. 

Suppose a user wants to find all experiments that have been processed by `tximport`. With the current DB schema, I can think of two approaches with pseudo-code like:
- Assume we have associated the tximport ComputationalResult record with each sample:
```Experiment.filter(samples[0].results.filter(command_executed.startwith('tximport ... <experiment_accession_code>).distinct()```

- Or: 
```
for each experiment:
    ComputationalResults.objects.filter(command_executed.startwith('tximport ... <experiment_accession_code>)
Then find the experiment based on <experiment_accession_code>
```
Neither of them is very satisfying to me.

dongbohu replied:
Of course, if this is not a valid user case, then we can just link this result with the samples to keep track of it w/o modifying the DB at all.

dongbohu replied:
Another question is: In the foreseeable future, is `tximport` going to be the only analysis that is at experiment level? If it is, then we can just treat as as a special case. If we will have other experiment-level jobs, then it is worth modifying the DB. Any thoughts, @jaclyn-taroni?

cgreene replied:
We have tried, to the extent possible, to select methods that work on single samples because it streamlines the processing greatly. I don't think I could guarantee that we won't have anything else, but we are trying to avoid it to the extent possible.

dongbohu replied:
Got it. I will associate this computational result record with all related samples for now. Although it may not help the query that much, at least it doesn't hurt. I will also add an issue to keep track of it.

kurtwheeler commented on workers/data_refinery_workers/processors/salmon.py:
```diff
@@ -373,6 +396,58 @@ def _run_fastqc(job_context: Dict) -> Dict:
+        error_template = ("Encountered error in R code while running tximport.R"
+                          " pipeline during processing of {0}: {1}")
+        error_message = error_template.format(exp_dir, str(e))
+        logger.error(error_message, processor_job=job_context["job_id"])
```
I think we should mark the job as a failure if this doesn't succeed. With the way this is right now, if this fails the only way we'll have any idea is if we see this error message in the logs.

dongbohu replied:
Sure. I can add a comment line about it.

dongbohu replied:
I didn't do it because even if `tximport` fails for whatever reason, the jobs that follows (salmontools, multiqc, etc) should still continue.

kurtwheeler replied:
I don't know that I agree that they should, however I'm not sure. This actually seems like a good question for @jaclyn-taroni: if `tximport` fails for some reason, would we prefer to have the rest of the pipeline continue or prevent that sample/experiment from being added to the compendium? It's not entirely clear to me how critical `tximport` is to have for a sample.

However, regardless of the answer to that question, I think we should probably fail the jobs at least for a while until we run into some errors that we can't find a workaround for, and then only let that kind of error through. The reason I think this is that as we run this on millions of samples, we're going to run into tons of edge cases which will cause failures. The way we're gonna debug these will be to look through our jobs/samples/experiments and figure out which didn't make it all the way through the system successfully and then track down what went wrong using the logs. We probably won't be going through every error message in the log files trying to figure out what happened, so if a job doesn't fail it won't make it onto our radar, and we won't realize that it failed for a fixable reason.

jaclyn-taroni replied:
> if tximport fails for some reason, would we prefer to have the rest of the pipeline continue or prevent that sample/experiment from being added to the compendium? It's not entirely clear to me how critical tximport is to have for a sample.

Hmm, this is a great question. I would say at this stage (Keytar Kurt) just fail the job, as right now (to my knowledge) the only things made available to the users are 1) count-scale output of `tximport` and 2) `multiqc` reports. In the future, we will probably want to allow users to obtain different outputs from the pipeline (e.g., transcript-level abundances from `salmon quant`, etc).

jaclyn-taroni replied:
Sounds good, thanks @dongbohu!

dongbohu commented:
@jaclyn-taroni @kurtwheeler and @Miserlou: I think I have addressed all your comments and fixed all conflicts with `dev` branch. Please let me know if you have other questions or concerns. Thanks.

jaclyn-taroni commented:
I have no additional comments

kurtwheeler commented:
Looks good to me, good work Dongbo!
